### DISCUSSION TITLE
How to use field resolver or extension for fields I want batched.

### DISCUSSION THREAD

**Discussion by jhirn** [rmosolgo/graphql-ruby#4806] [General]
Hello. I'm trying to figure out a way to write a custom field extension or resolver to use for basic cases when I want to batch load. In fact I'd love to automatically use this resolver/extension when the field is an ActiveRecord association, but one step at a time. 

I've implemented `Sources::Association` with data loader and can use it like this: 

```ruby
module Types
  class CountyType < Types::BaseObject
    field :cbsa, Types::CbsaType, null: false
    
    # repetative, yuck 
    def cbsa
      dataloader.with(Sources::Association, :cbsa).load(object)
    end
  end
end
```

But what I'd like is 
```ruby
module Types
  class CountyType < Types::BaseObject

    field :cbsa, Types::CbsaType, null: false, resolver: Resolvers::Association

    ## or 

    field :cbsa, Types::CbsaType, null: false, extensions: [Extensions::BatchAssociation]

    ## Ultimately (if possible)

    field :cbsa, Types::CbsaType, null: false ## determine object.field is AR association and automatically batch
  end
end
``` 

I'm trying to wrap my head around how to do this but it seemed difficult enough to convince me it might not be a great idea. In my head it feels like implementing resolve in an  Extension  would be simple enough but I'm not exactly sure how to get the field name to symbolize to pass to data loader. Something like 

```ruby
class AssociationExtension < GraphQL::Schema::FieldExtension
  def resolve(object:, arguments:, **rest)
      # naive, ignores arguments, could be enhanced. 
      dataloader.with(Sources::Association, FIELD_NAME).load(object)
   end
end
```

I suppose I could initialize it with the field name but it's redundant and verbose. 

```ruby
 field :cbsa, Types::CbsaType, null: false, extensions: [Extensions::BatchAssociation => {field_name: :cbsa]
```
Not exactly sure how the custom resolver would work as I don't want to declare args/type/etc... in the Resolver. I just want the resolve method to use my dataloader source. 

Anyways what are people's thoughts. Talk me out of this because it's spinning my wheels  🙃

**Comment by jhirn** [comment]
One possibility would be to use `context.current_path.last.to_sym`to get field name in the resolver but I'm not quite familiar enough with context to know if that's always going to be the case. 

```ruby
class AssociationExtension < GraphQL::Schema::FieldExtension
  def resolve(object:, arguments:, **rest)
      context.current_path #=> ["counties", 0, "cbsa"]
      field_name =  context.current_path.last.to_sym  #=> :cbsa
      dataloader.with(Sources::Association, field_name).load(object)
   end
end
```

Going to try this later on and report back, but also waiting for someone to talk me out of this idea entirely.

**Comment by rmosolgo** [comment]
FieldExtensions have a reference to the `field` they're attached to, that might help: https://github.com/rmosolgo/graphql-ruby/blob/930468fb4d39a62a21baf341f0c98932a15c3ba9/lib/graphql/schema/field_extension.rb#L12-L13

But I think if it was me, I'd use a resolver, something like this: 

```ruby 
class Types::BaseField < GraphQL::Schema::Field 
  class DataloadRelationResolver < GraphQL::Schema::Resolver 
    def resolve
      dataloader.with(Sources::Association, @field.original_name).load(object) 
    end 
  end 

  def initialize(*args, dataload_relation: false, **kwargs, &block)
    if dataload_relation 
      # TODO add a check to see if there's already a resolver, raise an ArgumentError if so
      kwargs[:resolver] = DataloadRelationResolver
    end 
    super(*args, **kwargs, &block)
  end 
end 
```

Then use it like: 

```ruby 
field :cbsa, Types::CbsaType, dataload_relation: true 
```

I wouldn't recommend using `context[:current_path]` because it will include any field aliases in the query, for example: 

```graphql 
{
  county(id: $id) {
    blah: cbsa { ... } 
    # ^^ in this case, context[:current_path] would have "blah" at the end
  }
}
```

**Comment by jhirn** [comment]
Very cool. I will give this a try.
